chain/chunks/src/lib.rs

```diff
@@ -1075,6 +1092,139 @@ impl ShardsManager {
         )
     }
 
+    /// Constructs a vector of [`ReceiptsProof`]s for specified chunk.
+    fn make_outgoing_receipts_proofs<T>(
+        &self,
+        chunk_header: &ShardChunkHeader,
+        outgoing_receipts: &Vec<Receipt>,
+        make_elem: impl Fn(ReceiptProof) -> T,
+    ) -> Result<Vec<T>, near_chunks_primitives::Error> {
+        let shard_id = chunk_header.shard_id();
+        let shard_layout = self
+            .runtime_adapter
+            .get_shard_layout_from_prev_block(chunk_header.prev_block_hash_ref())?;
+
+        let hashes = Chain::build_receipts_hashes(&outgoing_receipts, &shard_layout);
+        let (root, proofs) = merklize(&hashes);
+        if chunk_header.outgoing_receipts_root() != root {
+            return Err(near_chunks_primitives::Error::Other(format!(
+                "unexpected outgoing receipts root; want: {}, but got: {}",
+                chunk_header.outgoing_receipts_root(),
+                root
+            )));
+        }
+
+        let mut receipts_by_shard =
+            Chain::group_receipts_by_shard(outgoing_receipts.clone(), &shard_layout);
+        let proofs = proofs
+            .into_iter()
+            .enumerate()
+            .map(|(proof_shard_id, proof)| {
+                let proof_shard_id = proof_shard_id as u64;
+                let receipts = receipts_by_shard.remove(&proof_shard_id).unwrap_or_else(Vec::new);
+                let shard_proof = ShardProof {
+                    from_shard_id: shard_id,
+                    to_shard_id: proof_shard_id,
+                    proof: proof,
+                };
+                make_elem(ReceiptProof(receipts, shard_proof))
+            })
+            .collect();
+        Ok(proofs)
+    }
+
+    /// Prepares response to a partial encoded chunk request from a chunk read
+    /// from the storage.
+    ///
+    /// This requires encoding the chunk and as such is computationally
+    /// expensive operation.  If possible, the request should be served from
+    /// EncodedChunksCacheEntry or PartialEncodedChunk instead.
+    #[cfg_attr(feature = "test_features", visibility::make(pub))]
+    fn prepare_partial_encoded_chunk_response_from_chunk(
+        &mut self,
+        request: PartialEncodedChunkRequestMsg,
+        rs: &mut ReedSolomonWrapper,
+        chunk: ShardChunk,
+    ) -> Option<PartialEncodedChunkResponseMsg> {
+        let total_parts = rs.total_shard_count();
+        for ord in request.part_ords.iter() {
+            let ord: usize = (*ord).try_into().unwrap();
+            if ord >= total_parts {
+                warn!(target:"chunks", "Not sending {}, requested part_ord {} but we only expect {} total",
```

Comment:
This probably worth an `error`, right? Ie, this case signals a bug. (low level on confidence on this one)

Reply:
I honestly dunno. It’s a situation which does not prevent the node from operating (and for all we know the problem is on the other side) so that’s why I opted to using warning.

Reply:
Agree that warning is fine here because the request is sent from another node whose behavior we cannot control